### ISSUE TITLE
Memory Leak - RCTTiming + RCTConvert NSDate

### ISSUE THREAD

**Issue by ajwhite** [facebook/react-native#18042]
## TL;DR

`setTimeout` creates an `NSDate` allocation via `RCTConvert Date:` that never becomes released. **This appears to only happen in development mode**.

<img width="439" alt="screen shot 2018-02-21 at 11 39 38 am" src="https://user-images.githubusercontent.com/656630/36492796-f928f6e0-16fb-11e8-8b62-50f3bbc0d381.png">

### Is this a bug report?

Yes

### Have you read the [Contributing Guidelines](https://facebook.github.io/react-native/docs/contributing.html)?

Yes

### Environment

```
Environment:
  OS: macOS High Sierra 10.13.2
  Node: 6.10.2
  Yarn: 0.24.5
  npm: 3.10.10
  Watchman: 4.7.0
  Xcode: Xcode 9.2 Build version 9C40b
  Android Studio: 3.0 AI-171.4443003

Packages: (wanted => installed)
  react: 16.2.0 => 16.2.0
  react-native: 0.53.3 => 0.53.3
```

### Steps to Reproduce

`NSDate` objects become retained when creating timers in JS.

1. Create a timer
2. Run the app in development mode
3. Observe that `__NSDate` instances are retained

### Expected Behavior

`NSDate` objects become released upon the expiration of a timer.

### Actual Behavior

`NSDate` objects become retained when creating timers in JS.

<img width="1396" alt="screen shot 2018-02-21 at 11 34 33 am" src="https://user-images.githubusercontent.com/656630/36492463-37904740-16fb-11e8-9e83-573a287e2036.png">

### Reproducible Demo

https://github.com/ajwhite/react-native-memory-leak-nsdate

Create an endless timer and observe the memory growth. (https://github.com/ajwhite/react-native-memory-leak-nsdate/commit/c94205def099e5da01bedfe7f1bd4c0b14200ca7)

```js
// somewhere in your JS:

function timerLeak() {
  setTimeout(timerLeak, 100)
}

timerLeak()
```

Then open up Instruments and observe memory allocations under `__NSDate`
